Add granular mixing example (#551)
lethe-cfd/lethe: doc/source/examples/dem/granular-mixer/granular-mixer.rst

==================================
Granular Mixer
==================================

This example simulates the packing and mixing of particles in a mixer with a pitched-blade impeller. It is recommended to visit `DEM parameters <../../../parameters/dem/dem.html>`_ for more detailed information on the concepts and physical meanings of the parameters in Lethe-DEM.

Features
----------------------------------
- Solvers: ``dem_3d``
- Floating mesh
- Gmsh grids

> - Bidispersed particles

Reviewer: ```suggestion
- Bidispersed particle size
```

Reviewer: If they have the same size, I would instead specify something like particles with different ids or something like that.

Author: Added.

Location of the examples
------------------------
 ``/examples/dem/3d_granular_mixer/granular_mixer.prm``

Description of the case
-----------------------

> This simulation consists of two stages: packing (0-0.5 s) and mixing (0.5-5 s). There are two types of particles in this simulation (bidispersed system), that are inserted on top of each other during the packing stage. At :math:`t=0.5` s, the pitched-blade impeller starts rotating with an angular velocity of 6 rad/s and mixes the particles.

Reviewer: ```suggestion
This simulation consists of two stages: packing (0-0.5 s) and mixing (0.5-5 s) of particles. There are two types of particles in this simulation (with a respective diameter of X and Y), that are inserted on top of each other during the packing stage. At :math:`t=0.5` s, the pitched-blade impeller starts rotating with an angular velocity of 6 rad/s and mixes the particles.
```
I would specify the diameter of the particles directly

Reviewer: Ok now I just understood the have the same size. I would explain that in the context and make that clear from the get go that we have two particle type only to measure the mixing.

Author: Added to the case description.

Parameter file
--------------

Mesh
~~~~~

The background mesh (mixer body) is created using dealii ``subdivided_cylinder``.

.. code-block:: text

    subsection mesh
        set type                = dealii
        set grid type           = subdivided_cylinder
        set grid arguments      = 2 : 0.05 : 0.055
        set initial refinement  = 2
    end

Lagrangian physical properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are two types of particles with the same size and properties. We only need to define two particle types to make the visualization easier during post-processing.

.. code-block:: text

    subsection lagrangian physical properties
        set gx                                  = -9.81
        set gy                                  = 0.0
        set gz                                  = 0.0
        set number of particle types            = 2
            	subsection particle type 0
                set size distribution type              = uniform
                set diameter                            = 0.0015
                set number                              = 23500
                set density particles                   = 1500

>                 set young modulus particles         	  = 1000000

Reviewer: ```suggestion
                set young modulus particles         	  = 1e6
```
it's more clear when we use this notation.

                set poisson ratio particles             = 0.5
                set restitution coefficient particles   = 0.5
                set friction coefficient particles      = 0.5
                set rolling friction particles          = 0.2
        	end
        	subsection particle type 1
                set size distribution type              = uniform
                set diameter                            = 0.0015
                set number                              = 23500
                set density particles                   = 1500

>                 set young modulus particles             = 1000000

Reviewer: ```suggestion
                set young modulus particles             = 1e6
```

                set poisson ratio particles             = 0.5
                set restitution coefficient particles   = 0.5
                set friction coefficient particles      = 0.5
                set rolling friction particles          = 0.2
        	end

>                 set young modulus wall                  = 1000000

Reviewer: ```suggestion
                set young modulus wall                  = 1e6
```

                set poisson ratio wall                  = 0.5
                set restitution coefficient wall        = 0.5
                set friction coefficient wall           = 0.5
                set rolling friction wall               = 0.2
    end

Solid objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In this subsection, the floating meshes are defined. We can use dealii or gmsh to create the floating meshes. At the moment, solid objects (floating meshes) in Lethe have to be defined using triangular (simplex) meshes. Only triangular 2D meshes of 3D surfaces in the dem_3d solver are presently supported. Quadrilateral 2D meshes of 3D surfaces and 1D mesh of 2D surfaces are not supported at the moment. For each floating mesh, we need to specify a ``translational velocity``, a ``angular velocity``, and a ``center of rotation``. In this example, we only need an angular motion of the impeller. Note that the ``center of rotation`` of the impeller is at 0, 0, 0.

Reviewer 2: ```suggestion
In this subsection, the floating meshes are defined. We can use dealii or gmsh to create the floating meshes. At the moment, solid objects (floating meshes) in Lethe have to be defined using triangular (simplex) meshes. Only triangular 2D meshes of 3D surfaces in the dem_3d solver are presently supported. Quadrilateral 2D meshes of 3D surfaces and 1D mesh of 2D surfaces are not supported at the moment. For each floating mesh, we need to specify a ``translational velocity``, an ``angular velocity``, and a ``center of rotation``. In this example, we only need an angular motion of the impeller. Note that the ``center of rotation`` of the impeller is at 0, 0, 0.
```

.. code-block:: text

    subsection solid objects
        set number of solids        = 1
        subsection  solid object 0
        subsection mesh
            set type                = gmsh
            set file name           = pitched_blade_impeller.msh
            set simplex		       = true
            set initial refinement  = 0
        end
    	 
        subsection translational velocity
            set Function expression = 0 ; 0 ; 0
        end

        subsection angular velocity
            set Function expression = if(t > 0.5 , 6 ,0) ; 0 ; 0
        end
    
        subsection center of rotation
            # X COR
            set x				= 0
            # Y COR
            set y				= 0
            # Z COR
            set z				= 0
        end
    
      end
    
    end

Running the simulation
----------------------
This simulation can be launched by (in parallel mode on 16 processes):

.. code-block:: text

  mpirun -np 16 dem_3d granular_mixer.prm

.. warning::
	This example takes approximately 30 hours on 16 cores.

Results
---------

> Animation of the granular mixing simulation:

Reviewer: I would add a possibility for extension section where you could suggest that the mixing could be monitored using the distribution of the particle type. Otherwise that's a cool nice and simple example.

Author: Added.
